File: android_core_kit/src/main/java/vn/dylanbui/android_core_kit/mvvm_structure/ControllerLifecycleOwner.java
package vn.dylanbui.android_core_kit.mvvm_structure;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LifecycleRegistry;

import com.bluelinelabs.conductor.Controller;

/**
 * Created by IntelliJ IDEA.
 * User: Dylan Bui
 * Email: duc@propzy.com
 * Date: 6/8/20
 * To change this template use File | Settings | File and Code Templates.
 */



class ControllerLifecycleOwner implements LifecycleOwner, Controller.LifecycleListener {

    private final LifecycleRegistry lifecycleRegistry = new LifecycleRegistry(this);

    ControllerLifecycleOwner(Controller controller) {
        controller.addLifecycleListener(new Controller.LifecycleListener() {

            @Override
            public void preCreateView(@NonNull Controller controller) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE);
            }

            @Override
            public void postAttach(@NonNull Controller controller, @NonNull View view) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START);
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
            }

            @Override
            public void preDetach(@NonNull Controller controller, @NonNull View view) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE);
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
            }

            @Override
            public void postDestroyView(@NonNull Controller controller) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY);
            }

        });
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return lifecycleRegistry;
    }
}

File: android_core_kit/src/main/java/vn/dylanbui/android_core_kit/mvvm_structure/ControllerLifecycleOwner.kt
package vn.dylanbui.android_core_kit.mvvm_structure


/**
 * Created by IntelliJ IDEA.
 * User: Dylan Bui
 * Email: duc@propzy.com
 * Date: 5/29/20
 * To change this template use File | Settings | File and Code Templates.
 */

import android.view.View
import androidx.annotation.NonNull
import androidx.lifecycle.*
import com.bluelinelabs.conductor.Controller


class ControllerLifecycleOwner(controller: Controller) : Controller.LifecycleListener(), LifecycleOwner {
    private val lifecycleRegistry = LifecycleRegistry(this)

    @NonNull
    override fun getLifecycle(): Lifecycle {
        return lifecycleRegistry
    }

    init {
        controller.addLifecycleListener(object : Controller.LifecycleListener() {
            override fun preCreateView(@NonNull controller: Controller) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
            }

            override fun postAttach(controller: Controller, view: View) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
            }

            override fun preAttach(controller: Controller, view: View) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
            }

            override fun postDestroyView(@NonNull controller: Controller) {
                lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            }
        })
    }
}